streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -414,8 +417,11 @@ private void shutdownTasksAndState(final boolean rethrowExceptions) {
         log.debug("{} shutdownTasksAndState: shutting down all active tasks [{}] and standby tasks [{}]", logPrefix,
             activeTasks.keySet(), standbyTasks.keySet());
 
-        // Commit first as there may be cached records that have not been flushed yet.
-        commitOffsets(rethrowExceptions);
+        // only commit under clean exit
+        if (!caughtException) {
```

Comment:
I think there might be a bit more to it than this. Do we need to consider the flushing of the StateStores and the producer, too?
I.e, we have state in the state stores that is going to get flushed - that state is related to the offsets that we are not committing. That state is not-only going to be flushed to disk, but also written to the changelog topics. 
Although, without EOS i'm not sure we can do much. We can't guarantee that the state hasn't already been flushed to the store and the changelog, i.e., eviction can happen at any time.